### ISSUE TITLE
Upgrade to SecRel 4.0.0

### ISSUE THREAD

**Issue by sethdarragile6** [department-of-veterans-affairs/abd-vro#801]
**Description**

A release candidate for major version 4.0.0 (v4.0.0-rc1) has [been released](https://dsva.slack.com/archives/C02PTSZKM8W/p1670371063346419).

**Acceptance Criteria**

1. Upgrade to SecRel 4.0.0
2. Set up the new SecRel workflow as a PR check running <~8 minutes

**Notes about work**
- [Release Candidate notes](https://upgraded-invention-05777635.pages.github.io/pipeline/rc/)

**Comment by sethdarragile6** [comment]
We disabled the `security/snyk` step on all PRs after [this PR](https://github.com/department-of-veterans-affairs/abd-vro/pull/731/files#) was failing that step- even though the Aqua scan [passes](https://github.com/department-of-veterans-affairs/abd-vro-internal/actions/runs/3699570697). 

When similar discrepancies have happened in the past, it was explained by the SecRel team that the snyk.io scans operate purely on the source code, while the Aqua scans operate on the final build tree, i.e. that which will ultimately be deployed out. The latter is much more reliable and is in part what ultimately determines go/no-go for a release- however, it takes anywhere from 12-20 minutes to run.

SecRel 4.0.0 is supposed to dramatically reduce the amount of time it takes to run, making it much more feasible to run as a PR check. Per earlier discussions with @yoomlam, we'd like to bring this into the current sprint and make it happen. Yoom please correct me if I've misspoken here.

@ENeal49 

![image](https://user-images.githubusercontent.com/92405130/207743170-fc6fee0c-081b-41a7-b46e-22fbcc216cfd.png)

**Comment by yoomlam** [comment]
When trying to reference `env` within the `uses` block like this:
```yml
    uses: department-of-veterans-affairs/lighthouse-tornado-secrel-pipeline/.github/workflows/pipeline.yml@v4.0.0-rc1
    with:
      config-file: .github/secrel/config.yml
      images: |-
        ghcr.io/${{ github.repository }}/${{ env.TARGET_ENV }}_vro-app:${{ env.COMMIT_SHA }}
        ghcr.io/${{ github.repository }}/${{ env.TARGET_ENV }}_vro-postgres:${{ env.COMMIT_SHA }}
        ghcr.io/${{ github.repository }}/${{ env.TARGET_ENV }}_vro-db-init:${{ env.COMMIT_SHA }}
        ghcr.io/${{ github.repository }}/${{ env.TARGET_ENV }}_vro-console:${{ env.COMMIT_SHA }}
        ghcr.io/${{ github.repository }}/${{ env.TARGET_ENV }}_vro-service-data-access:${{ env.COMMIT_SHA }}
        ghcr.io/${{ github.repository }}/${{ env.TARGET_ENV }}_vro-service-pdfgenerator:${{ env.COMMIT_SHA }}
        ghcr.io/${{ github.repository }}/${{ env.TARGET_ENV }}_vro-service-assessclaimdc7101:${{ env.COMMIT_SHA }}
        ghcr.io/${{ github.repository }}/${{ env.TARGET_ENV }}_vro-service-assessclaimdc6602:${{ env.COMMIT_SHA }}
```
It's not possible due to [this limitation](https://docs.github.com/en/actions/using-workflows/reusing-workflows#limitations):
> Any environment variables set in an env context defined at the workflow level in the caller workflow are not propagated to the called workflow.

which is also repeated in [this constraint](https://docs.github.com/en/actions/learn-github-actions/contexts#env-context):
> You can use the env context in the value of any key in a step except for the id and uses keys.

cc: @ahay-agile6

**Comment by yoomlam** [comment]
@sethdarragile6 I got SecRel 4 to run [here](https://github.com/department-of-veterans-affairs/abd-vro-internal/actions/runs/3708485528) for branch `yoom/secrel4`, which is still a draft. However, at least you can see it run and the timing. I'll let you analyze it, clear some failures, and think about how we want to use it for PRs.

**Comment by yoomlam** [comment]
I'm thinking of leaving the current secrel.yml as it is and creating a separate secrel4prs.yml (SecRel for PRs) until we are comfortable with the new version.

**Comment by yoomlam** [comment]
@ahay-agile6 I'll take over this ticket and work with Seth, so you can focus on ramping up on DevOps.

**Comment by yoomlam** [comment]
I've done what I can for now. @sethdarragile6 will post some questions for the SecRel Team.
If we want to limit the frequency of SecRel runs (currently, it runs for every commit to all branches in the internal repo), we may can reduce how often the mirror action runs (easy) or implement another action (to listen for PR events on the public repo) to trigger the SecRel action (in the internal repo) -- I recommend the easier option, i.e., tweaking the mirror action, which means we'd have to update some wiki documentation.

**Comment by sethdarragile6** [comment]
On PR #860 @yoomlam  asked to provide a list of packages to be deleted (wait, why? see this [message](https://lighthouseva.slack.com/archives/C03UA9MV1EH/p1671643846200589?thread_ts=1671481495.084579&cid=C03UA9MV1EH) and thread). Capital idea! I wanted to lay those all out here explicitly, and get 👀 and buy-in from @advcoder + Jeremy and Stefan from SecRel. Along the way I have some clarifying questions.

### Packages to delete, then [manually update](https://animated-carnival-57b3e7f5.pages.github.io/starterkits/java/development-guide/#changing-published-package-visibility) when next they appear:
- [ ] abd-vro/vro-rabbitmq ([link](https://github.com/orgs/department-of-veterans-affairs/packages/container/package/abd-vro%2Fvro-rabbitmq)) - deploy to dev on public repo definitely created this
- [ ] dev_vro-app ([public](https://github.com/orgs/department-of-veterans-affairs/packages/container/package/abd-vro%2Fdev_vro-app))
- [ ] dev_vro-console  ([public](https://github.com/orgs/department-of-veterans-affairs/packages/container/package/abd-vro%2Fdev_vro-console))
- [ ] dev_vro-postgres  ([public](https://github.com/orgs/department-of-veterans-affairs/packages/container/package/abd-vro%2Fdev_vro-postgres))
- [ ] dev_vro-db-init  ([public](https://github.com/department-of-veterans-affairs/abd-vro/pkgs/container/abd-vro%2Fdev_vro-db-init))
- [ ] dev_vro-init-db ([public](https://github.com/department-of-veterans-affairs/abd-vro/pkgs/container/abd-vro%2Fdev_vro-init-db)) - probably disposable, no need to recreate
- [ ] dev_vro-service-pdfgenerator ([public](https://github.com/department-of-veterans-affairs/abd-vro/pkgs/container/abd-vro-internal%2Fdev_vro-service-pdfgenerator))
- [ ] dev_vro-service-assessclaimdc7101 ([public](https://github.com/department-of-veterans-affairs/abd-vro/pkgs/container/abd-vro-internal%2Fdev_vro-service-assessclaimdc7101))
- [ ] dev_vro-service-assessclaimdc6602 ([public](https://github.com/department-of-veterans-affairs/abd-vro/pkgs/container/abd-vro%2Fdev_vro-service-assessclaimdc6602))
- [ ] dev_vro-service-data-access ([public](https://github.com/orgs/department-of-veterans-affairs/packages/container/package/abd-vro%2Fdev_vro-service-data-access))
- [ ] qa_vro-app ([public](https://github.com/orgs/department-of-veterans-affairs/packages/container/package/abd-vro%2Fqa_vro-app))
- [ ] qa_vro-console  ([public](https://github.com/orgs/department-of-veterans-affairs/packages/container/package/abd-vro%2Fqa_vro-console))
- [ ] qa_vro-postgres  ([public](https://github.com/orgs/department-of-veterans-affairs/packages/container/package/abd-vro%2Fqa_vro-postgres))
- [ ] qa_vro-db-init  ([public](https://github.com/orgs/department-of-veterans-affairs/packages/container/package/abd-vro%2Fqa_vro-db-init))
- [ ] qa_vro-init-db ([public](https://github.com/orgs/department-of-veterans-affairs/packages/container/package/abd-vro%2Fqa_vro-init-db)) - probably disposable, no need to recreate
- [ ] qa_vro-service-pdfgenerator ([public](https://github.com/orgs/department-of-veterans-affairs/packages/container/package/abd-vro%2Fqa_vro-service-pdfgenerator))
- [ ] qa_vro-service-assessclaimdc7101 ([public](https://github.com/orgs/department-of-veterans-affairs/packages/container/package/abd-vro%2Fqa_vro-service-assessclaimdc7101))
- [ ] qa_vro-service-assessclaimdc6602 ([public](https://github.com/orgs/department-of-veterans-affairs/packages/container/package/abd-vro%2Fqa_vro-service-assessclaimdc6602))
- [ ] qa_vro-service-data-access ([public](https://github.com/orgs/department-of-veterans-affairs/packages/container/package/abd-vro%2Fqa_vro-service-data-access))
- [ ] sandbox_vro-app ([public](https://github.com/orgs/department-of-veterans-affairs/packages/container/package/abd-vro%2Fsandbox_vro-app))
- [ ] abd-vro/sandbox_vro-postgres ([public](https://github.com/orgs/department-of-veterans-affairs/packages/container/package/abd-vro%2Fsandbox_vro-postgres)) - probably disposable, no need to recreate
- [ ] sandbox_vro-db-init  ([public](https://github.com/orgs/department-of-veterans-affairs/packages/container/package/abd-vro%2Fsandbox_vro-db-init))
- [ ] sandbox_vro-service-pdfgenerator ([public](https://github.com/orgs/department-of-veterans-affairs/packages/container/package/abd-vro%2Fsandbox_vro-service-pdfgenerator))
- [ ] sandbox_vro-service-assessclaimdc7101 ([public](https://github.com/orgs/department-of-veterans-affairs/packages/container/package/abd-vro%2Fsandbox_vro-service-assessclaimdc7101))
- [ ] sandbox_vro-service-assessclaimdc6602 ([public](https://github.com/orgs/department-of-veterans-affairs/packages/container/package/abd-vro%2Fsandbox_vro-service-assessclaimdc6602))
- [ ] sandbox_vro-service-data-access ([public](https://github.com/orgs/department-of-veterans-affairs/packages/container/package/abd-vro%2Fsandbox_vro-service-data-access))

### Packages we may be able to ignore
- [ ] abd-vro-internal/vro-rabbitmq ([link](https://github.com/orgs/department-of-veterans-affairs/packages/container/package/abd-vro-internal%2Fvro-rabbitmq)) - was pushed by [this](https://github.com/department-of-veterans-affairs/abd-vro-internal/actions/runs/3754070499) from internal repo
- [ ] vro-rabbitmq ([link](https://github.com/department-of-veterans-affairs/lighthouse-secrel-test-fixtures/pkgs/container/lighthouse-secrel-test-fixtures%2Fvro-rabbitmq)) - may not be needed at all. Deploy to dev and secrel4prs should be pushing “abd-vro-internal/vro-rabbitmq” only. I had the “write issue” once, however, when every rabbitmq package was deleted but this one. I think this might’ve been BEFORE Jeremy fixed a different root cause (NOT the cross-posting) of the write issue, but I’m not 100% certain. Also, it has a README that suggests it’s not even ours, so… red herring?
- [ ] dev_vro-app ([internal](https://github.com/orgs/department-of-veterans-affairs/packages/container/package/abd-vro-internal%2Fdev_vro-app))
- [ ] dev_vro-console ([internal](https://github.com/orgs/department-of-veterans-affairs/packages/container/package/abd-vro-internal%2Fdev_vro-console))
- [ ] dev_vro-postgres ([internal](https://github.com/orgs/department-of-veterans-affairs/packages/container/package/abd-vro-internal%2Fdev_vro-postgres))
- [ ] dev_vro-db-init ([internal](https://github.com/orgs/department-of-veterans-affairs/packages/container/package/abd-vro-internal%2Fdev_vro-db-init))
- [ ] dev_vro-service-pdfgenerator ([internal](https://github.com/department-of-veterans-affairs/abd-vro-internal/pkgs/container/abd-vro-internal%2Fdev_vro-service-pdfgenerator))
- [ ] dev_vro-service-assessclaimdc7101 ([internal](https://github.com/department-of-veterans-affairs/abd-vro-internal/pkgs/container/abd-vro-internal%2Fdev_vro-service-assessclaimdc7101))
- [ ] dev_vro-service-assessclaimdc6602 ([internal](https://github.com/orgs/department-of-veterans-affairs/packages/container/package/abd-vro-internal%2Fdev_vro-service-assessclaimdc6602))
- [ ] dev_vro-service-data-access ([internal](https://github.com/orgs/department-of-veterans-affairs/packages/container/package/abd-vro-internal%2Fdev_vro-service-data-access))
- [ ] anything housed under `lighthouse-secrel-test-fixtures`, e.g. `dev_vro-console` below
![image](https://user-images.githubusercontent.com/92405130/209239922-35ceed14-d75a-4743-8df5-0f78cbf68472.png)

Need to confirm from Jeremy and Stefan that all the above is correct. I'm still a little uncertain whether or not the (internal) items in the Ignore List above are truly ignorable. I think they might also need to be deleted because `secrel4prs.yml` has been tagging and pushing via the PAT in `abd-vro-internal`, and the dockerfile LABELs are always associated to `abd-vro`. Once PR #860 is merged into develop, pushing will be solely in the context of GH token.

So the process as I see it looks like:
1. Merge in PR #860 
1. Wait until the dust settles on all GH actions triggered by that, noting the errors
1. Yoom, Ryan, or someone with package admin rights (not me, currently) deletes the above packages
1. Open a sample PR to `develop` with some minor change, or just a comment, merge it in
1. Wait until the dust settles on all GH actions triggered by that. Deploy to dev should kick off then, creating new packages in the GH token context
1. Package admin then [manually updates](https://animated-carnival-57b3e7f5.pages.github.io/starterkits/java/development-guide/#changing-published-package-visibility) all the newly created packages

Thoughts, comments, prayers? I'll probably copy paste a version of this whole spiel into the LH support channel, because I'm not sure if Jeremy and Stefan will be able to access this. Thanks

**Comment by jarzuaga-va** [comment]
I think you still want delete `abd-vro-internal/vro-rabbitmq` because it's living in the no mans land at the root of DVOA. The workflow link there was published by a branch using a PAT to login to the registry. If you try to re-publish this image using the same image name using the GITHUB_TOKEN from the internal repo, you're going to get a write permission denied error

**Comment by jarzuaga-va** [comment]
Step 3, you should just be able to re-run the failing pipeline after deletion of the packages and everything should publish correctly

**Comment by sethdarragile6** [comment]
> I think you still want delete `abd-vro-internal/vro-rabbitmq` because it's living in the no mans land at the root of DOVA. The workflow link there was published by a branch using a PAT to login to the registry. If you try to re-publish this image using the same image name using the GITHUB_TOKEN from the internal repo, you're going to get a write permission denied error

Interesting. I think that explains why I was still getting the write error even after Yoom deleted that package. Previously when you were talking about this "no man's land" item, I thought you were talking about [this](https://github.com/department-of-veterans-affairs/lighthouse-secrel-test-fixtures/pkgs/container/lighthouse-secrel-test-fixtures%2Fvro-rabbitmq) (which is truly bizarre). That suggests to me that we're going to have to make more changes around how we specifically push `rabbitmq`...

docker pull rabbitmq:3
docker tag rabbitmq:3 "ghcr.io/${REPO}/vro-rabbitmq:3"
docker push "ghcr.io/${REPO}/vro-rabbitmq:3"

**Comment by yoomlam** [comment]
Deleted the following:
- abd-vro/vro-rabbitmq
- abd-vro/dev_vro-app
- abd-vro/dev_vro-console
- abd-vro/dev_vro-postgres
- abd-vro/dev_vro-db-init - why is this the case?: "probably disposable, no need to recreate" 
- abd-vro-internal/dev_vro-service-pdfgenerator
- abd-vro-internal/dev_vro-service-assessclaimdc7101
- abd-vro/dev_vro-service-assessclaimdc6602
- abd-vro/dev_vro-service-data-access
- abd-vro-internal/vro-rabbitmq

Skipping `qa_` and `sandbox_` until we are sure and understand how this will work in the future.
I don't have permission to delete:
- abd-vro/sandbox_vro-postgres ([public](https://github.com/orgs/department-of-veterans-affairs/packages/container/package/abd-vro%2Fsandbox_vro-postgres))

Who will be able to delete the new packages?

**Comment by sethdarragile6** [comment]
> Deleted the following:
> * abd-vro/dev_vro-db-init - why is this the case?: "probably disposable, no need to recreate"

There's a "abd-vro/dev_vro-**db**-init" (the one we use) and a "abd-vro/dev_vro-**init**-db", which is the one I was referring to. Looks to me like an initial version of the package before someone decided they like **db-init** better than **init-db**. Not critical to address now, I just listed it as something we might want to clean up eventually.

> 
> Who will be able to delete the new packages?

Unknown. 😬 I might run a little experiment and push up a test copy of one of the packages as a sanity check. Jeremy might have some insight into that, since they have been building in that context all along  @jarzuaga-va

**Comment by sethdarragile6** [comment]
After working through issues with the SecRel team over break, we finally got to a successful [full pass](https://github.com/department-of-veterans-affairs/abd-vro-internal/actions/runs/3833361500) today. In the end, I wound up only needing to delete and regenerate the rabbitmq, pdfgenerator and the assessclaimdc7101 packages. 

Of course now, version 4.0.0-**rc2** is has been [released](https://lighthouseva.slack.com/archives/C03UA9MV1EH/p1672771142066839) (this effort was all for 4.0.0-rc1). I'm going to move this ticket to Done, however, and start a new ticket that would involve:

1. upgrading to 4.0.0-rc2
2. more testing/tweaking of secrel4prs.yml generally
3. replacing secrel.yml with secrel4prs.yml, making it our "defacto" secrel scan

@yoomlam @jarzuaga-va @dianagriffin

**Comment by sethdarragile6** [comment]
closing out in favor of #888
